CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,513 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import boomerang.BackwardQuery;
+import boomerang.Boomerang;
+import boomerang.DefaultBoomerangOptions;
+import boomerang.ForwardQuery;
+import boomerang.callgraph.ObservableDynamicICFG;
+import boomerang.callgraph.ObservableICFG;
+import boomerang.jimple.Statement;
+import boomerang.jimple.Val;
+import boomerang.preanalysis.BoomerangPretransformer;
+import boomerang.results.AbstractBoomerangResults;
+import boomerang.results.BackwardBoomerangResults;
+import boomerang.seedfactory.SeedFactory;
+import crypto.rules.CryptSLRule;
+import crypto.rules.CryptSLRuleReader;
+import soot.Body;
+import soot.G;
+import soot.PackManager;
+import soot.Scene;
+import soot.SceneTransformer;
+import soot.SootClass;
+import soot.SootMethod;
+import soot.Transform;
+import soot.Transformer;
+import soot.Type;
+import soot.Unit;
+import soot.Value;
+import soot.jimple.Stmt;
+import soot.jimple.TableSwitchStmt;
+import soot.jimple.internal.JAssignStmt;
+import soot.jimple.internal.JIfStmt;
+import soot.jimple.internal.JStaticInvokeExpr;
+import soot.options.Options;
+import wpds.impl.Weight.NoWeight;
+
+public class ProviderDetection {
+	
+	private String provider = null;
+	private String rulesDirectory = null;
+	private String defaultRulesDirectory = System.getProperty("user.dir")+File.separator+"src"+File.separator+"main"+File.separator+"resources"+File.separator+"JavaCryptographicArchitecture";
+	
+	private String sootClassPath = System.getProperty("user.dir") + File.separator+"target"+File.separator+"classes";
+	private String sootClassPathForTests = System.getProperty("user.dir") + File.separator+"target"+File.separator+"test-classes";
+	
+	public String getProvider() {
+		return provider;
+	}
+
+	public String getRulesDirectory() {
+		return rulesDirectory;
+	}
+	
+	
+	/**
+	 * This method is used to get the Soot classpath from `src/main/java`
+	 */
+	public String getMainSootClassPath() {
+		//Assume target folder to be directly in user dir
+		File classPathDir = new File(sootClassPath);
+		if (!classPathDir.exists()){
+			throw new RuntimeException("Classpath for the test cases could not be found.");
+		}
+		return sootClassPath;
+	}
+	
+	/**
+	 * This method is used to get the Soot classpath from `src/test/java`
+	 * in order to run the JUnit test cases for Provider Detection
+	 */
+	public String getSootClassPath(){
+		//Assume target folder to be directly in user dir
+		File classPathDir = new File(sootClassPathForTests);
+		if (!classPathDir.exists()){
+			throw new RuntimeException("Classpath for the test cases could not be found.");
+		}
+		return sootClassPathForTests;
+	}
+	
+	/**
+	 * This method is used to setup Soot
+	 */
+	public void setupSoot(String sootClassPath, String mainClass) {
```

Comment:
What is the logic of your provider detection?

We are running Soot twice. Once for Provider Detection and afterwards again for the general analysis. Is that intentional? Could we circumvent it? This is not the most performant solution, as setting up soot and constructing the call graph may take up to a couple of minutes, dependening on the project we analyse.

Reply:
I have added this `setupSoot()` method when running the provider detection test cases only. It is not used when running the actual provider detection analysis. During the general analysis + provider analysis, Soot is run only once.

Reply:
@johspaeth in the `HeadlessCryptoScanner.java` only this method [here](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/50943e06214cc009ce7c368c94c28c6db0919c4b/CryptoAnalysis/src/main/java/crypto/HeadlessCryptoScanner.java#L256) from `ProviderDetection.java` is called. The implementation of that method can be found [here](https://github.com/CROSSINGTUD/CryptoAnalysis/blob/50943e06214cc009ce7c368c94c28c6db0919c4b/CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java#L172) and other methods from `ProviderDetection.java` like setupSoot or createAnalysisTransformer are not called from HeadlessCryptoScanner, since they are already setup there. These methods are only used when running the provider detection test cases, as briefly explained above.